### ISSUE TITLE
More involved parsing

### ISSUE THREAD

**Issue by jonhoo** [mattnenterprise/rust-imap#28]
Currently, all the commands simply return `Vec<String>`. It'd be great if they returned more semantically relevant results (i.e., fully parsed structs). This would also let us make the IDLE handle return an `Iterator` over said structs, so you could monitor changes to the mailbox in real-time.

**Comment by jonhoo** [comment]
This also applies to arguments. For example, the `sequence_set` argument to `fetch` should probably be a `Vec<Uid>` or something, and should probably return something akin to a `HashMap<Uid, _>`.

**Comment by sanmai-NL** [comment]
Would you be interested to cooperate with @djc on this? He has written [another IMAP crate](https://crates.io/crates/tokio-imap) that supports [`nom`-based server response parsing](https://github.com/djc/tokio-imap/blob/20721abc6c756c3522e0266a0954728cead13935/src/parser.rs). I suppose a concerted effort is the best way forward. 🙂

**Comment by djc** [comment]
If you're interested we can probably split the protocol data structures and parser out of tokio-imap so that they can more easily be reused. In terms of encoding, so far I've mostly used str slices, although I recently changed the parser so that it returns bytes for message contents (for the RFC822 FETCH attribute only so far).

**Comment by jonhoo** [comment]
@djc That would be awesome!

**Comment by djc** [comment]
The imap-proto crate is now live on crates.io: https://crates.io/crates/imap-proto.

**Comment by jonhoo** [comment]
Docs at https://docs.rs/imap-proto/0.1.0/imap_proto/. Not too wordy atm, but still useful.
I don't know that I'll have time to look into incorporating this particularly soon, but it's on my radar. Otherwise, PRs welcome!

**Comment by djc** [comment]
I probably won't work on the integration between rust-imap and imap-proto, but if people find gaps in imap-proto for stuff that rust-imap, I can have a look at those.

**Comment by jonhoo** [comment]
@djc yeah, I wasn't expecting that you would, just wanted to comment for others watching this issue :)
I might contribute some updates to the docs for `imap-proto` when I get around to incorporating it too.

**Comment by jonhoo** [comment]
@djc from a cursory review, it looks like `imap-proto` expects all strings to be valid UTF-8. As has been pointed out elsewhere (see #54 for a summary), this can be a problem for some applications. Do you have a plan for the future for how you'll deal with servers that use other charsets?

**Comment by jonhoo** [comment]
Filed djc/imap-proto#5 including a proposed solution.
